README.md
# txt2page

[txt2page](https://github.com/ds26gte/txt2page) is a bash/vim/groff script that converts
free-form plain text into HTML and PDF. groff is needed only for PDF.

```
% txt2page filename.txt
````

creates `filename.html`.  Here, txt2page is assumed already placed
in your `PATH`, and `%` is your Un⋆x command-line prompt.

[List of commands possible.](./cheatsheet.txt)

You can supply groff-like options to txt2page. For HTML, the only
relevant one is `-s`.

For PDF, use the option `-Tpdf` and other usual groff options for
[fonts](./fontsetup/otfgroff.md), layout, &c. E.g.

```
% txt2page -Tpdf -r PS=14p -r PI=1.5m -r LL=6i -r PO=1.13i filename.txt
````

For HTML, if the `-s` option is not used,

```
 .so filename Some Additional Description
````

merely ⎈links⎈ to (the HTML version of) `filename` with
`Some Additional Description` serving as link text.  In troff of
course, both syntaxes cause sourcing of `filename`.  I had it
this way because it’s a way to coax a Table of Contents into the
HTML without additional markup.

If txt2page is called with the `-s` option, however, all `.so`
calls cause interpolation (recursively).  This allows you to
create one single all-encompassing HTML page for the document,
chasing down all subfiles.


history
2015 Nov 27

Some adoc-isms: link:, -, url[...], ----, =

2015 Nov 26

Use .REDIRECT <url> to redirect (as in troff2page).

2015 Sept 21

Distinguish -man and -ms .SH

2015 June 1

\*[urlh URL]LINK TEXT\& becomes <a href="URL">LINK TEXT</a>, as
in troff2page. (For troff, \*[urlh ...] is defined to be an empty
string, so the whole expression becomes just LINK TEXT.)

\& is empty string, as in troff[2page].

Automatically recognized filenames don't have [].

.=* is title.

2015 May 31

Added option: -c css_file

2015 Mar 20

.= etc for troff-like headers that allow heading to spill across
lines.

2015 Mar 1

Determine title-text after headers have all been determined, so
troff-like headers can contribute a potential multiple-line
title.

2014 Dec 16

‡ text URL

for urlh.

2014 Dec 14

Use ⚓ for invisible anchor; ※ for visible anchor (reference).

2014 Dec 5

᛫ text URL

(spaces optional) gets converted to <a href="URL">text</a>

2014 Nov 2

txt2page, if given multiple file arguments, creates HTML files
for each of them, and adds previous/next nav bars to each. Got
rid of explicit .NAV command as not needed.

For -Tpdf, for single file, output pdf has same basename as that
file. Otherwise, it's txt2pageout.pdf.

2014 Oct 29

HTML looks for <basename>.css, txt2page.css, default.css,
<anyother>.css.  PDF looks for <basename>.tmac, txt2page.tmac,
default.tmac.  Latter can be used to modify .TITLE and .SEC
macros.

2014 Oct 25

As in troff, flush-left line should continue previous line. (Only
non-flush-left lines cause a linebreak of previous line.)

2014 Oct 22

Any lines with only non-alnums that is preceded and followed by
blank lines is treated as a section-breaking ornament.

2014 Oct 21

Use leading :: to introduce obeylines, which ends on a blank
line. No need for any other closer.

Empty level-3 section with 3 trailing #s (### ###) introduces a
drop cap.

(Motivation: Recycle the # section syntax to introduce these
sectioning variations, so they will be ignored gracefully by other
#-recognizers.)

Retire txt2page.tmac.

Obeylines env should begin/end with implicit noindent.

Use a dummy space for a blank line within code display so lynx
doesn't swallow it.

2014 Oct 20

Cleaner DC annotation.

2014 Oct 13

.nf and ``` should preserve contiguous spaces even if they aren't
leading

2014 Oct 6

Use :: to fence obeylines

2014 Oct 1

Make treatment of &<> more maintainable: save all Æ«» (if any) as
@tzp_...@ strings; convert all source &<> to Æ«»; do the
HTMLization; all &<> are now valid HTML; convert Æ«» to
&amp;&lt;&gt; ; then finally restore the original Æ«»

2014 Sep 26

*=REDIRECT=* URL

2014 Sep 19

Allow ``` to be preceded by spaces

2014 Sep 17

URLs shouldn't contain brackets.

2014 Sep 12

Recognize :emoji:.

2014 Aug 22

Clean up function Clean_up_urls -- shouldn't have space before >
in <a href="...">

2014 Aug 21

Call ex with -u NONE

2014 Jun 25

Use HTML5 doctype

2014 Jun 16

... shouldn't cause lisphilite to choke

< > & should be verbatimized in subfiles (.so'd files) too

2014 Apr 11

& inside URLs should not be HTMLized

2013 Sep 9

# ... ## produces just HTML title, nothing in text

2013 Apr 1

Shunt pdf part into a subscript txt2page_pdf

2013 Mar 31

| tables also translated for pdf

2013 Mar 23

# introduces doc title if single # closes it.  Otherwise regular
section title.

2013 Mar 19

Insert a &nbsp; for a blank line within a listing; this is only
for lynx, which compresses successive <br/>s into one.

2013 Mar 18

lisphilite.rkt

2013 Mar 15

lisphilite.lisp a standalone filter, doesn't depend on tex2page.

2013 Mar 13

'^\*\s' for top-level bulleted grafs.

2013 Mar 12

For an input file f.txt tack on any f__*.txt files that exist.
This is to add navigation-bar info without hardcoding it into the
main file. (Of course you can put the nav bar in the main file
too if you want.  But I'd like to be able to rearrange chapters
without modifying the main text.) Use _ as placeholder in navbar
for greyed-out parts (e.g., for first and last page).

Allow *...* URL to be split across (2 or 3?) lines.

※ (changed 2014 Dec 14) creates visible anchors.

2013 Mar 11

⚓ (changed 2014 Dec 14) to create (invisible) anchors (HTML node names).  Refer to them as
./#anchor for anchor in own file, and ./anotherfile#anchor for
anchors in another file.

Make sure that hrefs convert from source
extension to .html.  Source document can keep source extension to
facilitate file navigaton (vim gf) entirely in source.

Leading .'s inside ``` are automatically escaped.

Added lisphilite.sh (which depends on tex2page for now).
Highlights any annotated ``` as lisp. Any additional lisp symbols
to be recognized as keywords should be put in ./.lispwords.

syntaxhl.css.

2013 Mar 10

Added GFM-style sections (with leading & possible trailing #'s) and
listings (```).

``...`` -- this allows ` to be enclosed.

2013 Feb 17

Convert some -'s to minus signs. E.g. 3 - 2, -2, ->, but not
123-4567 or 980-123-4567.

2012 Nov 16

Incorporated PDF conversion capability into txt2page.  The -Tpdf
option creates a PDF page

2012 Nov 14

txt2pdffile: a leading period that isn't part of a command gets
an implicit \& prefix

txtpage: -s option (previously -soelim): so named to share syntax
with groff.  With -s, any .so files within the document is
immediately inserted (and recursively), instead of being treated
as a link to another HTML file (presumably created with another
call to txt2page).  Thus, one comprehensive HTML output page is
created.

txt2page: don't allow .txt extension of argument filename to be
optional.  This is less TeX-like but more troff-like

Simplify footnote mark syntax: ^^[sym]

Allow \ (except for \"comment) to show up as itself in PDF

2012 Nov 9

Corrected indentation and lineskip errors caused if very first
line is indented.

2012 Sep 28

Links to files with same extension as source are treated as links
to corresponding HTML files and displayed as such.

2012 Sep 7


lisphilite.rkt
#! /usr/bin/env racket

#lang racket

;dorai.sitaram@gmail.com
;
;first change 2013-03-15
;last change 2013-03-17

;just the syntax-highlighting code out of tex2page

(define *hilite-token-delims*
  (list #\( #\) #\[ #\] #\{ #\} #\' #\` #\" #\; #\, #\|))

(define *hilite-keywords*
  (list

    "=>"
    "and"
    "begin" "begin0" "block"
    "case" "cond"
    "define" "define-syntax" "delay" "do"
    "else"
    "if"
    "lambda" "let" "let*" "let-syntax" "let-values" "letrec" "letrec-syntax"
    "or"
    "quasiquote" "quote"
    "set!" "syntax-case" "syntax-rules"
    "unless" "unquote" "unquote-splicing" "unwind-protect"
    "when" "with-handlers" 

    ))

(define (hilite-read-lispwords)
  (let ((f ".lispwords"))
    (when (file-exists? f)
      (call-with-input-file
        f (lambda (i)
            (let loop ()
              (let ((c (peek-char i)))
                (unless (eof-object? c)
                  (cond ((char=? c #\;) (read-line i))
                        ((or (char-whitespace? c)
                             (ormap (lambda (x) (char=? x c))
                                    *hilite-token-delims*))
                         (read-char i))
                        (else (set! *hilite-keywords*
                                (cons (list->string
                                        (reverse
                                          (let loop ((s '()) (esc-p #f))
                                            (let ((c (peek-char i)))
                                              (cond ((eof-object? c) s)
                                                    (esc-p (loop (cons (read-char i) s) #f))
                                                    ((char=? c #\\ ) (loop (cons (read-char i) s) #t))
                                                    ((or (char-whitespace? c)
                                                         (ormap
                                                           (lambda (x)
                                                             (char=? c x))
                                                           *hilite-token-delims*))
                                                     s)
                                                    (else (loop (cons (read-char i) s)
                                                                #f)))))))
                                      *hilite-keywords*))))
                  (loop)))))))))

(define (hilite-output-token s)
  (printf "<span class=~a>~a</span>"
          (hilite-get-type s) s))

(define (hilite-get-type s)
  (cond ((ormap (lambda (x) (string=? s x)) *hilite-keywords*)
         'keyword)
        ((char=? (string-ref s 0) #\#) 'selfeval)
        ((string->number s) 'selfeval)
        (else 'variable)))

(define (hilite-write-char c)
  (display (case c
             ((#\<) "&lt;")
             ((#\>) "&gt;")
             ((#\") "&quot;")
             ((#\&) "&amp;")
             (else c))))

(define (hilite-output-string)
  (read-char)
  (display "<span class=selfeval>")
  (hilite-write-char #\")
  (let loop ((esc-p #f))
    (let ((c (read-char)))
      (case c
        ((#\") (when esc-p
                 (hilite-write-char c)
                 (loop #f)))
        ((#\\ ) (write-char c)
                (loop #t))
        (else (hilite-write-char c)
              (loop #f)))))
  (hilite-write-char #\")
  (display "</span>"))

(define (hilite-output-comment)
  (display "<span class=comment>")
  (let loop ()
    (let ((c (read-char)))
      (cond ((or (eof-object? c)
                 (char=? c #\newline)
                 (and (char=? c #\return)
                      (let ((c (peek-char)))
                        (or (eof-object? c)
                            (char=? c #\newline)
                            (begin (read-char) #t)))))
             #t)
            (else (hilite-write-char c)
                  (loop)))))
  (printf "</span>~%"))

(define (hilite-output-next-chunk)
  (let ((c (peek-char)))
    (unless (eof-object? c)
      (cond ((char=? c #\;) (hilite-output-comment))
            ((char=? c #\") (hilite-output-string))
            ((or (char=? c #\') (char=? c #\`) (char=? c #\,))
             (read-char)
             (printf "<span class=keyword>~c" c)
             (when (and (char=? c #\,) (char=? (peek-char) #\@))
               (write-char (read-char)))
             (printf "</span>"))
            ((or (char-whitespace? c)
                 (ormap (lambda (x) (char=? x c)) *hilite-token-delims*))
             (hilite-write-char (read-char)))
            (else
              (hilite-output-token (hilite-get-token)))))))

(define (hilite)
  (let loop ()
    (let ((c (peek-char)))
      (unless (eof-object? c)
        (hilite-output-next-chunk)
        (loop)))))

(define (hilite-get-token)
  (list->string
    (reverse
      (let loop ((s '()) (esc-p #f))
        (let ((c (peek-char)))
          (cond ((eof-object? c) s)
                (esc-p (loop (cons (read-char) s) #f))
                ((char=? c #\\ ) (loop (cons (read-char) s) #t))
                ((or (char-whitespace? c)
                     (ormap (lambda (x) (char=? c x)) *hilite-token-delims*))
                 s)
                (else (loop (cons (read-char) s) #f))))))))

(hilite-read-lispwords)

(define hilite-test
  (lambda (f)
    (with-input-from-file f hilite)))

(hilite)


lisphilite.lisp
":";if test "$LISP" = clisp; then exec clisp -q $0
":";elif test "$LISP" = clozure; then exec ccl -b -Q -l $0
":";elif test "$LISP" = ecl; then exec ecl -shell $0
":";else exec sbcl --script $0
":";fi

;Dorai Sitaram

;just the syntax-highlighting code out of tex2page

;first change 2013-03-15
;last change 2015-05-29

(defvar *lisphi-token-delims*
  (list #\( #\) #\[ #\] #\{ #\} #\' #\` #\" #\; #\, #\|))

(defvar *it*)

(defun lisphi-string-to-number (s)
  (if (position #\: s :test #'char=) nil
    (let ((n (read-from-string s nil)))
      (if (numberp n) n nil))))

(defun lisphi-char-whitespace-p (c)
  (or (char= c #\space)
      (char= c #\tab)
      (not (graphic-char-p c))))

(defvar *lisphi-keywords*
  (list

    "=>"
    "..."
    "and"
    "assert"
    "begin"
    "begin0"
    "block"
    "case"
    "cond"
    "decf"
    "define"
    "define-macro"
    "define-syntax"
    "defmacro"
    "defpackage"
    "defparameter"
    "defstruct"
    "defun"
    "defvar"
    "delay"
    "destructuring-bind"
    "do"
    "do-all-symbols"
    "do-external-symbols"
    "do-symbols"
    "dolist"
    "dotimes"
    "ecase"
    "else"
    "etypecase"
    "eval-when"
    "flet"
    "fluid-let"
    "handler-bind"
    "handler-case"
    "if"
    "incf"
    "labels"
    "lambda"
    "let"
    "let*"
    "let-syntax"
    "let-values"
    "letrec"
    "letrec-syntax"
    "loop" ;NB! common variable in Scheme
    "macrolet"
    "multiple-value-bind"
    "multiple-value-setq"
    "or"
    "pop"
    "prog1"
    "progn"
    "push"
    "quasiquote"
    "quote"
    "set!"
    "setf"
    "setq"
    "syntax-case"
    "syntax-rules"
    "typecase"
    "unless"
    "unquote"
    "unquote-splicing"
    "unwind-protect"
    "when"
    "with"
    "with-handlers"
    "with-input-from-string"
    "with-open-file"
    "with-open-socket"
    "with-open-stream"
    "with-output-to-string"
    "with-slots"

    ))

(defun lisphi-read-lispwords ()
  (let ((f ".lispwords"))
    (when (probe-file f)
      (with-open-file (i f :direction :input)
        (loop
          (let ((c (peek-char nil i nil)))
            (cond ((not c) (return))
                  ((or (lisphi-char-whitespace-p c)
                       (member c *lisphi-token-delims* :test #'char=))
                   (read-char i))
                  (t (push
                       (concatenate
                         'string
                         (nreverse
                           (let ((s '()) (esc-p nil))
                             (loop
                               (let ((c (peek-char nil i nil)))
                                 (cond ((not c) (return s))
                                       (esc-p (read-char i)
                                              (push c s)
                                              (setq esc-p nil))
                                       ((char= c #\\ )
                                        (read-char i)
                                        (push c s)
                                        (setq esc-p t))
                                       ((or (lisphi-char-whitespace-p c)
                                            (member c *lisphi-token-delims* :test #'char=))
                                        (return s))
                                       (t (read-char i)
                                          (push c s))))))))
                       *lisphi-keywords*)))))))))

(defun lisphi-get-type (s)
  (let (it)
    (cond ((member s *lisphi-keywords* :test 'string=) :keyword)
          ((setq it (position #\: s :test #'char=))
           (if (= it 0) :selfeval :variable))
          ((char= (char s 0) #\#) :selfeval)
          ((lisphi-string-to-number s) :selfeval)
          (t :variable))))

(defun lisphi-princ-char (c)
  (when c
    (cond ((char= c #\newline) (terpri))
          (t (princ (case c
                      ((#\<) "&lt;")
                      ((#\>) "&gt;")
                      ((#\") "&quot;")
                      ((#\&) "&amp;")
                      (t c)))))))

(defun lisphi-output-string ()
  (read-char)
  (princ "<span class=selfeval>")
  (lisphi-princ-char #\")
  (let ((esc-p nil))
    (loop
      (let ((c (read-char nil nil)))
        (case c
          (#\" (unless esc-p (return))
           (lisphi-princ-char c)
           (setq esc-p nil))
          (#\\ (princ c)
           (setq esc-p (not esc-p)))
          (t (lisphi-princ-char c)
             (setq esc-p nil))))))
  (lisphi-princ-char #\")
  (princ "</span>"))

(defun lisphi-output-comment ()
  (princ "<span class=comment>")
  (loop
    (let ((c (read-char nil nil)))
      (cond ((or (not c) (char= c #\newline)
                 (and (lisphi-char-whitespace-p c)
                      (let ((c2 (peek-char nil nil nil)))
                        (or (not c2) (char= c2 #\newline)))))
             (princ "</span>")
             (terpri)
             (return))
            (t (lisphi-princ-char c))))))

(defun lisphi-get-token ()
  (concatenate
    'string
    (nreverse
      (let ((s '()) (esc-p nil))
        (loop
          (let ((c (peek-char nil nil nil)))
            (when (not c) (return s))
            (cond (esc-p (read-char)
                         (push c s)
                         (setq esc-p nil))
                  ((char= c #\\ )
                   (read-char)
                   (push c s)
                   (setq esc-p t))
                  ((or (lisphi-char-whitespace-p c)
                       (member c *lisphi-token-delims* :test #'char=))
                   (return s))
                  (t (read-char)
                     (push c s)))))))))

(defun lisphi-princ (s)
  (let ((n (length s)) (k 0))
    (loop
      (unless (< k n) (return))
      (lisphi-princ-char (char s k))
      (incf k))))

(defun lisphi-output-token (s)
  (case (lisphi-get-type s)
    (:keyword
      (princ "<span class=keyword>")
      (lisphi-princ s)
      (princ "</span>"))
    (:selfeval
      (princ "<span class=selfeval>")
      (lisphi-princ s)
      (princ "</span>"))
    (t (princ "<span class=variable>")
       (lisphi-princ s)
       (princ "</span>"))))

(defun lisphi-output-next-chunk ()
  (let ((c (peek-char nil nil nil)))
    (cond ((char= c #\;)
           (lisphi-output-comment))
          ((char= c #\")
           (lisphi-output-string))
          ((or (char= c #\') (char= c #\`) (char= c #\,))
           (read-char)
           (princ "<span class=keyword>")
           (princ c)
           (when (char= c #\,)
             (let ((c (peek-char nil nil nil)))
               (when (char= c #\@)
                 (read-char)
                 (princ c))))
           (princ "</span>"))
          ((or (lisphi-char-whitespace-p c)
               (member c *lisphi-token-delims* :test #'char=))
           (read-char)
           (lisphi-princ-char c))
          (t (lisphi-output-token
               (lisphi-get-token))))))

(defun lisphi ()
  (loop
    (let ((c (peek-char nil nil nil)))
      (unless c (return)))
    (lisphi-output-next-chunk)))

(lisphi-read-lispwords)

(defun lisphi-test (f)
  (with-open-file (*standard-input* f :direction :input)
    (lisphi)))

(lisphi)
